File: 프로그래머스/3/214288. 상담원 인원/상담원 인원.py
from collections import defaultdict
import heapq
def split(n,k):
    result = []
    def dfs(path, rest_s, rest_k):
        # path = 지금까지 고른 배열
        # rest_s = 남은 합
        # rest_k = 남은 k
        if rest_k == 0:
            if rest_s == 0:
                result.append(path[:])
            return
        min_v = 1
        max_v = rest_s - (rest_k-1)
        if max_v < min_v: return
        for value in range(min_v, max_v+1):
            path.append(value)
            dfs(path, rest_s - value, rest_k-1)
            path.pop()
    dfs([],n,k)
    return result
def solution(k, n, reqs):
    ans = float('inf')
    # 기다린 시간의 합이 최소
    # n명의 멘토를 k로 최소 몫 1이상 나누기
    # (시작, 끝)으로 넣고
    # if len(stack) == s[c] and stack[-1]: 기다리는 시간 계산해서 더해주기
    for s in split(n,k):
        w = 0
        s = [0] + s
        counsling = defaultdict(list)
        for i in range(1,k+1):
            # 각 상담 유형마다 개수에 맞는 상담사 heap을 생성(끝나는 시간 기준)
            for _ in range(s[i]):
                heapq.heappush(counsling[i],0)
        for a,b,c in reqs:
            end = heapq.heappop(counsling[c])
            if end > a:
                w += end - a
                a = end
            heapq.heappush(counsling[c],a+b)
        ans = min(ans,w)
    return ans

File: 프로그래머스/3/214288. 상담원 인원/상담원 인원.java
import java.util.*;
class Solution {
    public int solution(int k, int n, int[][] reqs) {
        int ans = Integer.MAX_VALUE;
        List<List<Integer>> sp = split(n,k);
        for (List<Integer> s : sp) {
            HashMap<Integer,PriorityQueue<Integer>> cousling = new HashMap<>();
            for (int idx = 1; idx <= k; idx++) {
                // 각 상담 유형(k-1)마다 우선순위 큐 생성 후 미리 배치된 상담사만큼 초기화
                PriorityQueue<Integer> pq = new PriorityQueue<>();
                int cmp = s.get(idx-1);
                for (int i = 0; i < cmp; i++) {
                    pq.offer(0);
                }
                cousling.put(idx,pq);
            }
            int wait = 0;
            for (int[] r : reqs) {
                int a = r[0];
                int b = r[1];
                int c = r[2];
                Integer end = cousling.get(c).poll();
                if (end > a) {
                    wait += end - a;
                    a = end;
                }
                cousling.get(c).offer(a+b);
            }
            ans = Math.min(ans,wait);
        }
        return ans;
    }
    
    public List<List<Integer>> split(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(new ArrayList<Integer>(),n,k,result);
        return result;
    }
    
    public void dfs(List<Integer> path, int rest_sum, int rest_k, List<List<Integer>> result) {
        if (rest_k == 0) {
            if (rest_sum == 0) result.add(new ArrayList<>(path));
            return;
        }
        
        Integer min = 1;
        Integer max = rest_sum - (rest_k-1);
        if (min > max) return;
        for (int value = min; value <= max; value++) {
            path.add(value);
            dfs(path,rest_sum-value,rest_k-1,result);
            path.remove(path.size() - 1);
        }
    }
}